### PR TITLE
decode variable with mismatched coordinate attribute

### LINKED ISSUE #1809: WRF output : cannot serialize variable
#### Code Sample, a copy-pastable example if possible

```python
md.to_netcdf('modified_wrf_input.nc')

---------------------------------------------------------------------------
ValueError                                Traceback (most recent call last)
<ipython-input-55-ccba003701f9> in <module>()
----> 1 md.to_netcdf('modified_wrf_input.nc')

/Users/gbromley/anaconda/lib/python3.5/site-packages/xarray/core/dataset.py in to_netcdf(self, path, mode, format, group, engine, encoding, unlimited_dims)
    981         return to_netcdf(self, path, mode, format=format, group=group,
    982                          engine=engine, encoding=encoding,
--> 983                          unlimited_dims=unlimited_dims)
    984 
    985     def __unicode__(self):

/Users/gbromley/anaconda/lib/python3.5/site-packages/xarray/backends/api.py in to_netcdf(dataset, path_or_file, mode, format, group, engine, writer, encoding, unlimited_dims)
    581     try:
    582         dataset.dump_to_store(store, sync=sync, encoding=encoding,
--> 583                               unlimited_dims=unlimited_dims)
    584         if path_or_file is None:
    585             return target.getvalue()

/Users/gbromley/anaconda/lib/python3.5/site-packages/xarray/core/dataset.py in dump_to_store(self, store, encoder, sync, encoding, unlimited_dims)
    907         if encoding is None:
    908             encoding = {}
--> 909         variables, attrs = conventions.encode_dataset_coordinates(self)
    910 
    911         check_encoding = set()

/Users/gbromley/anaconda/lib/python3.5/site-packages/xarray/conventions.py in encode_dataset_coordinates(dataset)
   1052     non_dim_coord_names = set(dataset.coords) - set(dataset.dims)
   1053     return _encode_coordinates(dataset._variables, dataset.attrs,
-> 1054                                non_dim_coord_names=non_dim_coord_names)
   1055 
   1056 

/Users/gbromley/anaconda/lib/python3.5/site-packages/xarray/conventions.py in _encode_coordinates(variables, attributes, non_dim_coord_names)
   1015             raise ValueError('cannot serialize coordinates because variable '
   1016                              "%s already has an attribute 'coordinates'"
-> 1017                              % var_name)
   1018         attrs['coordinates'] = ' '.join(map(str, coord_names))
   1019 

ValueError: cannot serialize coordinates because variable FSA already has an attribute 'coordinates'

```
#### Problem description

This dataset is the wrfinput_d01 file for use with the WRF model. Reading and modifying the variables worked, but I cannot figure out how to write the changes out to the file. I get the above error. I saw another post about the same issue, but wasn't sure how it was resolved.

### LINKED ISSUE #5170: to_netcdf is not idempotent when stacking rename and set_coords
After doing
```python
# !wget https://storage.ecmwf.europeanweather.cloud/s2s-ai-challenge/dev/xarray-issue/source.nc
ds = xr.open_dataset('source.nc') 
ds = ds.rename({'number': 'n'})
ds = ds.set_coords('valid_time')
print(ds)
ds.to_netcdf('out.nc')
ds = xr.open_dataset('out.nc')
print(ds)
```
`'valid_time'` is not a coordinate in the netcdf file coordinate is turned into a variable.

```
<xarray.Dataset>
Dimensions:            (heightAboveGround: 1, latitude: 121, longitude: 240, n: 2, step: 3, time: 1)
Coordinates:
  * n                  (n) int64 0 1
  * time               (time) datetime64[ns] 2020-01-02
  * step               (step) timedelta64[ns] 1 days 2 days 3 days
  * heightAboveGround  (heightAboveGround) int64 2
  * latitude           (latitude) float64 90.0 88.5 87.0 ... -87.0 -88.5 -90.0
  * longitude          (longitude) float64 0.0 1.5 3.0 4.5 ... 355.5 357.0 358.5
    valid_time         (time, step) datetime64[ns] ...
Data variables:
    t2m                (n, time, step, heightAboveGround, latitude, longitude) float32 ...
Attributes:
    GRIB_edition:            2
    GRIB_centre:             ecmf
    GRIB_centreDescription:  European Centre for Medium-Range Weather Forecasts
    GRIB_subCentre:          0
    Conventions:             CF-1.7
    institution:             European Centre for Medium-Range Weather Forecasts
    history:                 2021-04-15T12:08:59 GRIB to CDM+CF via cfgrib-0....
```

```
<xarray.Dataset>
Dimensions:            (heightAboveGround: 1, latitude: 121, longitude: 240, n: 2, step: 3, time: 1)
Coordinates:
  * n                  (n) int64 0 1
  * time               (time) datetime64[ns] 2020-01-02
  * step               (step) timedelta64[ns] 1 days 2 days 3 days
  * heightAboveGround  (heightAboveGround) int64 2
  * latitude           (latitude) float64 90.0 88.5 87.0 ... -87.0 -88.5 -90.0
  * longitude          (longitude) float64 0.0 1.5 3.0 4.5 ... 355.5 357.0 358.5
Data variables:
    valid_time         (time, step) datetime64[ns] ...
    t2m                (n, time, step, heightAboveGround, latitude, longitude) float32 ...
Attributes:
    GRIB_edition:            2
    GRIB_centre:             ecmf
    GRIB_centreDescription:  European Centre for Medium-Range Weather Forecasts
    GRIB_subCentre:          0
    Conventions:             CF-1.7
    institution:             European Centre for Medium-Range Weather Forecasts
    history:                 2021-04-15T12:08:59 GRIB to CDM+CF via cfgrib-0....
```

**What happened**:
Output of the MCVE is :
Coords  written in copy.1.nc: ['number', 'time', 'step', 'heightAboveGround', 'latitude', 'longitude', 'valid_time']
Reread and check   copy.1.nc: ['number', 'time', 'step', 'heightAboveGround', 'latitude', 'longitude', 'valid_time']

Coords  written in copy.2.nc: ['n', 'time', 'step', 'heightAboveGround', 'latitude', 'longitude', 'valid_time']
Reread and check   copy.2.nc: ['n', 'time', 'step', 'heightAboveGround', 'latitude', 'longitude']

Coords  written in copy.4.nc: ['n', 'time', 'step', 'heightAboveGround', 'latitude', 'longitude', 'valid_time']
Reread and check   copy.4.nc: ['n', 'time', 'step', 'heightAboveGround', 'latitude', 'longitude', 'valid_time']

**What you expected to happen**
Output of the MCVE should be :

Coords  written in copy.1.nc: ['number', 'time', 'step', 'heightAboveGround', 'latitude', 'longitude', 'valid_time']
Reread and check   copy.1.nc: ['number', 'time', 'step', 'heightAboveGround', 'latitude', 'longitude', 'valid_time']

Coords  written in copy.2.nc: ['n', 'time', 'step', 'heightAboveGround', 'latitude', 'longitude', 'valid_time']
Reread and check   copy.2.nc: ['n', 'time', 'step', 'heightAboveGround', 'latitude', 'longitude'**, 'valid_time'**]

Coords  written in copy.4.nc: ['n', 'time', 'step', 'heightAboveGround', 'latitude', 'longitude', 'valid_time']
Reread and check   copy.4.nc: ['n', 'time', 'step', 'heightAboveGround', 'latitude', 'longitude', 'valid_time']

**Minimal Complete Verifiable Example**:

```python

import xarray as xr
import numpy as np

xr.show_versions()
print(xr.__file__)

FILE1 = 'copy.1.nc'
FILE2 = 'copy.2.nc'
FILE3 = 'copy.3.nc'
FILE4 = 'copy.4.nc'

ds = xr.open_dataset('source.nc')

# Initial dataset is ok
print(f'Coords  written in {FILE1}: {list(ds.coords)}')
ds.to_netcdf(FILE1)
print(f'Reread and check   {FILE1}: {list(xr.open_dataset(FILE1).coords)}')
#print(xr.open_dataset(FILE1))

print()

# No round trip
ds = ds.rename({'number': 'n'})
ds = ds.set_coords('valid_time')
print(f'Coords  written in {FILE2}: {list(ds.coords)}')
ds.to_netcdf(FILE2)
print(f'Reread and check   {FILE2}: {list(xr.open_dataset(FILE2).coords)}')
#print(xr.open_dataset(FILE2))

print()

# Doing a round trip solves the issue :
ds = xr.open_dataset('source.nc')
ds = ds.rename({'number': 'n'})
ds.to_netcdf(FILE3)
ds = xr.open_dataset(FILE3)
ds = ds.set_coords('valid_time')
print(f'Coords  written in {FILE4}: {list(ds.coords)}')
ds.to_netcdf(FILE4)
print(f'Reread and check   {FILE4}: {list(xr.open_dataset(FILE4).coords)}')
#print(xr.open_dataset(FILE4))
```

**Anything else we need to know?**:
The initial source.nc file is available at  wget https://storage.ecmwf.europeanweather.cloud/s2s-ai-challenge/dev/xarray-issue/source.nc

**Environment**:

<details><summary>Output of <tt>xr.show_versions()</tt></summary>

INSTALLED VERSIONS
------------------
commit: None
python: 3.8.6 | packaged by conda-forge | (default, Jan 25 2021, 23:21:18)
[GCC 9.3.0]
python-bits: 64
OS: Linux
OS-release: 3.10.0-1160.15.2.el7.x86_64
machine: x86_64
processor: x86_64
byteorder: little
LC_ALL: None
LANG: en_US.UTF-8
LOCALE: en_US.UTF-8
libhdf5: 1.12.0
libnetcdf: 4.7.4

xarray: 0.17.0
pandas: 1.2.2
numpy: 1.19.5
scipy: None
netCDF4: 1.5.6
pydap: None
h5netcdf: None
h5py: 2.10.0
Nio: None
zarr: 2.6.1
cftime: 1.4.1
nc_time_axis: None
PseudoNetCDF: None
rasterio: None
cfgrib: 0.9.8.5
iris: None
bottleneck: None
dask: 2021.02.0
distributed: 2021.02.0
matplotlib: 3.3.4
cartopy: None
seaborn: None
numbagg: None
pint: None
setuptools: 49.6.0.post20210108
pip: 21.0.1
conda: 4.9.2
pytest: 5.3.1
IPython: 7.20.0
sphinx: 3.4.3

</details>

It may be related/duplicate of #4512 and linked to https://github.com/pydata/xarray/pull/4108

### PR DESCRIPTION
**Pull request by kmuehlbauer**
- decode variable with mismatched coordinate attribute
- propagate `coordinates` to encoding in any case

- [x] Closes #1809
- [x] Closes #5170 
- [x] Tests added
- [x] User visible changes (including notable bug fixes) are documented in `whats-new.rst`
- [ ] New functions/methods are listed in `api.rst`

### CODE REVIEW

**Comment by kmuehlbauer** [comment]
This is still draft, because I wanted to get more insight on the approach first.

For now this will properly decode and propagate `coordinates` to `encoding`, but the missing coordinate is removed from that string. Additionally a warning is raised. I've added a "coordinates_strict" to the available options for `decode_coords` to enable a workflow which raises in the case of mismatched coordinates.

Some questions:

- Should we just not propagate `coordinate` to encoding in case of mismatches?
- Should we also check if `coordinates` are matching on write (see #5170)?

Any comments appreciated!

File: xarray/conventions.py [hunk 1]
```diff
@@ -438,15 +438,31 @@ def stackable(dim: Hashable) -> bool:
             )
         except Exception as e:
             raise type(e)(f"Failed to decode variable {k!r}: {e}")
-        if decode_coords in [True, "coordinates", "all"]:
+        if decode_coords in [True, "coordinates", "all", "coordinates_strict"]:
```
Comment (Illviljan) [inline xarray/conventions.py:441]: This implies `decode_coords: Literal[True, "coordinates", "all", "coordinates_strict"]`, but the argument says `bool`-only. 
How come mypy hasn't complained about this before?

File: xarray/conventions.py [hunk 2]
```diff
@@ -438,15 +438,31 @@ def stackable(dim: Hashable) -> bool:
...
-                coord_str = var_attrs["coordinates"]
-                var_coord_names = coord_str.split()
-                if all(k in variables for k in var_coord_names):
-                    new_vars[k].encoding["coordinates"] = coord_str
-                    del var_attrs["coordinates"]
-                    coord_names.update(var_coord_names)
+                var_coord_names = var_attrs["coordinates"].split()
+                existing_coord_names = [c for c in var_coord_names if c in variables]
+                missing_coord_names = [c for c in var_coord_names if c not in variables]
+                if missing_coord_names:
```
Comment (Illviljan) [inline xarray/conventions.py:447]:
```suggestion
                existing_coord_names = [c for c in var_coord_names if c in variables]
                if not existing_coord_names:
                    missing_coord_names = [c for c in var_coord_names if c not in variables]
```
`missing_coord_names` looks like a error state calculation, we don't want to run that calculation if everything works as intended.

I wonder if these comparison can be done using `set`-comparison?

File: xarray/conventions.py [hunk 3]
```diff
@@ -438,15 +438,31 @@ def stackable(dim: Hashable) -> bool:
...
+                        from xarray.core.utils import emit_user_level_warning
+
+                        emit_user_level_warning(
+                            " ".join(
+                                [
+                                    msg,
+                                    f"Decoding with new ``coordinates``: {existing_coord_names}.",
+                                ]
+                            )
+                        )
```
Comment (Illviljan) [inline xarray/conventions.py:461]:
```suggestion
                       msg_solution = f"Decoding with new ``coordinates``: {existing_coord_names}."
                        emit_user_level_warning(" ".join([msg, msg_solution]))
```

File: xarray/conventions.py (hunk 1 above)
Comment (kmuehlbauer) [reply xarray/conventions.py:441]: reveal_type(decode_coords) gives: 

`xarray/conventions.py:442: note: Revealed type is "builtins.bool"`

Comment (Illviljan) [reply xarray/conventions.py:441]: yep, which must've been wrong considering possible values in main was `"coordinates"` and `"all"`.

Replace the input type with `decode_coords: Literal[True, "coordinates", "all", "coordinates_strict"]`.

File: xarray/conventions.py (hunk 2 above)
Comment (kmuehlbauer) [reply xarray/conventions.py]: > I wonder if these comparison can be done using `set`-comparison?

Yes, that greatly simplified the workflow.

**Comment by kmuehlbauer** [comment]
This solves the problem in #1809.

But I'm not really happy with the naming of `decode_coords`-strings.

**Comment by kmuehlbauer** [comment]
OK, thanks to the test suite we now have a contradicting issue here:

https://github.com/pydata/xarray/issues/308

@shoyer It would be great if you can share your thoughts on how to handle mismatched coordinates here. In https://github.com/pydata/xarray/pull/313 it silently does not handle mismatched coordinates and keeps mismatched coordinates attributes on the variable. Should we just raise on mismatch (no good idea!) or keep current behaviour but issue warning (might be overlooked or disabled)? Or is there some other solution, beside the proposed by this PR?

**Comment by dcherian** [comment]
I think we shouldn't add more options and just silently ignore the names of any missing variables. This attribute gets out of sync quite easily.

> Should we just not propagate coordinate to encoding in case of mismatches?

Let's propagate if `decode_coords is not False`

> Should we also check if coordinates are matching on write

We could warn if the coordinates attribute refers to absent variables at write time. But I suspect it would just be annoying.

**Comment by kmuehlbauer** [comment]
Thanks @dcherian, that greatly simplifies the code.

As a compromise with #308 coordinates are now only propagated to encoding if at least one coordinate name matches a variable.

**Comment by kmuehlbauer** [comment]
As this now does not emit any warning, should I revert the change using `warnings.warn` -> `emit_user_level_warning` or is this OK to proceed?

File: xarray/backends/api.py [hunk 4]
```diff
@@ -488,6 +488,10 @@ def open_dataset(
           as coordinate variables.
         - "all": Set variables referred to in  ``'grid_mapping'``, ``'bounds'`` and
           other attributes as coordinate variables.
+
+        Only existing variables can be set as coordinates. Missing variables
```
Comment (kmuehlbauer) [inline xarray/backends/api.py:492]: I hope that this is short enough but explains how missing variables are treated. No native speaker here, so please suggest better wording/grammar.

File: doc/whats-new.rst [hunk 5]
```diff
@@ -84,6 +84,8 @@ Bug fixes
   issues (:issue:`7817`, :issue:`7942`, :issue:`7790`, :issue:`6191`, :issue:`7096`,
   :issue:`1064`, :pull:`7827`).
   By `Kai Mühlbauer <https://github.com/kmuehlbauer>`_.
+- Fixed a bug where mismatched ``coordinates`` silently failed to decode variable (:issue:`1809`, :pull:`8195`).
```
Comment (dcherian) [inline doc/whats-new.rst:87]:
```suggestion
- Fixed a bug where inaccurate ``coordinates`` silently failed to decode variable (:issue:`1809`, :pull:`8195`).
```

File: xarray/backends/api.py [hunk 6]
```diff
@@ -488,6 +488,10 @@ def open_dataset(
...
+        will be silently ignored. If ``coordinates`` do not match any variable,
+        no decoding takes place.
```
Comment (dcherian) [inline xarray/backends/api.py:494]:
```suggestion
        will be silently ignored. 
```

File: xarray/backends/api.py [hunk 7]
```diff
@@ -691,6 +695,10 @@ def open_dataarray(
...
```
Comment (dcherian) [inline xarray/backends/api.py:701]:
```suggestion
        will be silently ignored.
```

File: xarray/conventions.py [hunk 8]
```diff
@@ -514,14 +514,17 @@ def decode_cf(
     decode_times : bool, optional
         Decode cf times (e.g., integers since "hours since 2000-01-01") to
         np.datetime64.
-    decode_coords : bool or {"coordinates", "all"}, optional
+    decode_coords : bool or {"coordinates", "coordinates_strict", "all", "all_strict"}, optional
```
Comment (dcherian) [inline xarray/conventions.py:517]:
```suggestion
    decode_coords : bool or {"coordinates", "all"}, optional
```

File: xarray/conventions.py [hunk 9]
```diff
@@ -514,14 +514,17 @@ def decode_cf(
...
         Controls which variables are set as coordinate variables:
 
-        - "coordinates" or True: Set variables referred to in the
+        - "coordinates"/"coordinates_strict" or True: Set variables referred to in the
           ``'coordinates'`` attribute of the datasets or individual variables
           as coordinate variables.
-        - "all": Set variables referred to in  ``'grid_mapping'``, ``'bounds'`` and
+        - "all"/"all_strict": Set variables referred to in  ``'grid_mapping'``, ``'bounds'`` and
```
Comment (dcherian) [inline xarray/conventions.py:523]:
```suggestion
        - "all": Set variables referred to in  ``'grid_mapping'``, ``'bounds'`` and
```

File: xarray/conventions.py [hunk 10]
```diff
@@ -514,14 +514,17 @@ def decode_cf(
...
           other attributes as coordinate variables.
+        When using the "strict"-forms an error is raised, if ``coordinates``-items are
+        missing from the dataset, otherwise a warning is issued and only the
+        available ``coordinates`items are handled.
```
Comment (dcherian) [inline xarray/conventions.py:527]:
```suggestion
```

File: xarray/tests/test_conventions.py [hunk 11]
```diff
@@ -80,6 +80,28 @@ def test_decode_cf_with_conflicting_fill_missing_value() -> None:
...
+    orig = Dataset(
+        {
+            "XLONG": (["x", "y"], zeros1.squeeze(0), {}),
+            "XLAT": (["x", "y"], zeros1.squeeze(0), {}),
+            "foo": (["time", "x", "y"], zeros1, {"coordinates": "XTIME XLONG XLAT"}),
+            "time": ("time", [0.0], {"units": "hours since 2017-01-01"}),
+        }
+    )
+    decoded = conventions.decode_cf(orig, decode_coords=True)
+    assert decoded["foo"].encoding["coordinates"] == "XLONG XLAT"
```
Comment (dcherian) [inline xarray/tests/test_conventions.py:96]:
```suggestion
    assert decoded["foo"].encoding["coordinates"] == "XTIME XLONG XLAT"
```

File: xarray/tests/test_conventions.py [hunk 12]
```diff
@@ -80,6 +80,28 @@ def test_decode_cf_with_conflicting_fill_missing_value() -> None:
...
+    assert list(decoded.coords.keys()) == ["XLONG", "XLAT", "time"]
+
+    decoded = conventions.decode_cf(orig, decode_coords=False)
+    assert decoded["foo"].encoding.get("coordinates") is None
```
Comment (dcherian) [inline xarray/tests/test_conventions.py:100]:
```suggestion
    assert "coordinates" not in decoded["foo"].encoding
```

File: xarray/conventions.py [hunk 13]
```diff
@@ -441,10 +440,12 @@ def stackable(dim: Hashable) -> bool:
...
-                coord_str = var_attrs["coordinates"]
-                var_coord_names = coord_str.split()
-                if all(k in variables for k in var_coord_names):
-                    new_vars[k].encoding["coordinates"] = coord_str
+                var_coord_names = [
+                    c for c in var_attrs["coordinates"].split() if c in variables
+                ]
+                # only decode if there is at least one variable matching a coordinate
+                if var_coord_names:
+                    new_vars[k].encoding["coordinates"] = " ".join(var_coord_names)
```
[the remark this answers is not in the document]
Comment (kmuehlbauer) [reply xarray/conventions.py]: OK, I thought, that might spoil roundtripping.

File: xarray/conventions.py (hunk 10 above)
Comment (kmuehlbauer) [reply xarray/conventions.py]: Thanks @dcherian for catching this. I should not commit shortly before leaving.

**Review by dcherian — APPROVED** [review]
LGTM :) but could use a pass from someone else to double check our decisions

**Comment by kmuehlbauer** [comment]
> LGTM :) but could use a pass from someone else to double check our decisions

Current behaviour:

`coordinates` are propagated to `encoding` if all coordinates are in variables

New behaviour:

`coordinates` are propagated to `encoding` if at least one coordinate is in variables

File: xarray/conventions.py [hunk 14]
```diff
@@ -441,10 +440,12 @@ def stackable(dim: Hashable) -> bool:
...
-                var_coord_names = coord_str.split()
-                if all(k in variables for k in var_coord_names):
-                    new_vars[k].encoding["coordinates"] = coord_str
+                var_coord_names = [
+                    c for c in var_attrs["coordinates"].split() if c in variables
+                ]
+                # only decode if there is at least one variable matching a coordinate
+                if var_coord_names:
+                    new_vars[k].encoding["coordinates"] = var_attrs["coordinates"]
                     del var_attrs["coordinates"]
```
Comment (dcherian) [inline xarray/conventions.py:449]: Actually i think I prefer this. We always move "coordinates" over if `decode_coords is not False`. THat way it is predictable.

```suggestion
                new_vars[k].encoding["coordinates"] = var_attrs["coordinates"]
                del var_attrs["coordinates"]
                if var_coord_names:
```

Comment (kmuehlbauer) [reply xarray/conventions.py:449]: That means: `coordinates` are propagated to `encoding` if given in the source file. Yes, that's even simpler.

Comment (kmuehlbauer) [reply xarray/conventions.py]: Oh, that one test need fixing.

**Comment by kmuehlbauer** [comment]
Current behaviour (if `decode_coords` is not `False`):

`coordinates` are only propagated to `encoding` if all coordinates are in variables 

New behaviour (if `decode_coords` is not `False`):

`coordinates` are propagated to `encoding` in any case, but only coordinates in variables are added as coords

**Pull request merged**